pattern/equivalence/AlphaEquivalence.java

```diff
@@ -17,262 +17,115 @@
-    public abstract boolean isValid();
-
-    public abstract Valid asValid();
-
-    public abstract AlphaEquivalence addOrInvalidate(Supplier<AlphaEquivalence> mappingSupplier);
-
-    protected abstract AlphaEquivalence addOrInvalidate(AlphaEquivalence mapping);
-
-    static AlphaEquivalence create(Map<Variable, Variable> map) {
-        Map<Variable, Variable> reverseMap = new HashMap<>();
-        for (Map.Entry<Variable, Variable> e : map.entrySet()) {
-            reverseMap.put(e.getValue(), e.getKey());
-        }
-        if (map.size() != reverseMap.size()) return new AlphaEquivalence.Invalid();
-        return new AlphaEquivalence.Valid(map, reverseMap);
+    public static AlphaEquivalence empty() {
+        return new AlphaEquivalence(new HashMap<>(), new HashMap<>());
     }
 
-    public static class Valid extends AlphaEquivalence {
-
-        private Map<Variable, Variable> map;
-        private Map<Variable, Variable> reverseMap;
-
-        private Valid(Map<Variable, Variable> map, Map<Variable, Variable> reverseMap) {
-            assert map.size() == reverseMap.size();
-            assert map.keySet().equals(set(reverseMap.values()));
-            this.map = map;
-            this.reverseMap = reverseMap;
-        }
+    public FunctionalIterator<AlphaEquivalence> alphaEqualIf(boolean condition) {
+        if (condition) return Iterators.single(this);
+        else return Iterators.empty();
+    }
 
-        public static AlphaEquivalence.Valid create() {
-            return new AlphaEquivalence.Valid(new HashMap<>(), new HashMap<>());
+    public FunctionalIterator<AlphaEquivalence> extendIfCompatible(AlphaEquivalence alphaMap) {
+        Optional<Map<Variable, Variable>> m = mergeVariableMapping(variableMapping(), alphaMap.variableMapping());
+        Optional<Map<Variable, Variable>> r = mergeVariableMapping(reverseVariableMapping(), alphaMap.reverseVariableMapping());
+        if (m.isPresent() && r.isPresent()) {
+            if (m.get().size() != r.get().size()) return Iterators.empty();
+            return Iterators.single(new AlphaEquivalence(m.get(), r.get()));
+        } else {
+            return Iterators.empty();
         }
+    }
 
-        @Override
-        public AlphaEquivalence addOrInvalidate(AlphaEquivalence alphaMap) {
-            if (!alphaMap.isValid()) return AlphaEquivalence.invalid();
-            Optional<Map<Variable, Variable>> m = mergeVariableMapping(variableMapping(), alphaMap.asValid().variableMapping());
-            Optional<Map<Variable, Variable>> r = mergeVariableMapping(reverseVariableMapping(), alphaMap.asValid().reverseVariableMapping());
-            if (m.isPresent() && r.isPresent()) {
-                map = m.get();
-                reverseMap = r.get();
-                if (map.size() != reverseMap.size()) return AlphaEquivalence.invalid();
-                return this;
+    private static Optional<Map<Variable, Variable>> mergeVariableMapping(Map<Variable, Variable> existing,
```

Comment:
modifying an input is sometimes the right style to use (generally we don't unless we have to). When we do, we should then not return that modified object back because the API looks like it has created a new object.

This function is particularly bad because it can modify the `existing` variable for a while, then fail and return an empty Optional. The API reads as if the operation failed and nothing happened, but we've implicitly updated the mapping as well.

Reply:
There's another smell to chase here which is that `variableMapping()` always returns a new copy of the `map` that this class holds, which is why this wasn't broken before.

It's probably cleaner to have separate APIs for `mapping()` and `mappingCopy()` and use them as appropriate - best to check the places `variableMapping()` is called in this class once over.

Reply:
> There's another smell to chase here which is that `variableMapping()` always returns a new copy of the `map` that this class holds, which is why this wasn't broken before.

Ah yes I do this where I deem appropriate because it's in the Essential Java book which advocates it to guard against exactly the kind of bug I wrote here. I realise we aren't doing this throughout the codebase though, so better to stick to the style we have. I've checked and we can have just one getter and make a copy where needed.